### PR TITLE
FRONT-2449: Button Pattern.

### LINKED ISSUE #16: Refactor Button pattern with BCL component

### PR DESCRIPTION
**Pull request by tibi2303**

### CODE REVIEW

**Review by drishu — CHANGES_REQUESTED** [review]
see todo in templates/overrides/form/input--submit.html.twig
this needs to be implemented.
Take notice of button attributes (primary secondary) and change according to the pattern variants.

Please add `- "bcl"` under `file_scan_ignore_directories` in runner.dist.yml until https://github.com/openeuropa/oe_bootstrap_theme/issues/61 is solved.

**Comment by escuriola** [comment]
> see todo in templates/overrides/form/input--submit.html.twig
> this needs to be implemented.
> Take notice of button attributes (primary secondary) and change according to the pattern variants.
> 
> Please add `- "bcl"` under `file_scan_ignore_directories` in runner.dist.yml until #61 is solved.

This would print a button which by itself that not submit the form.

**Comment by drishu** [comment]
> > see todo in templates/overrides/form/input--submit.html.twig
> > this needs to be implemented.
> > Take notice of button attributes (primary secondary) and change according to the pattern variants.
> > Please add `- "bcl"` under `file_scan_ignore_directories` in runner.dist.yml until #61 is solved.
> 
> This would print a button which by itself that not submit the form.

I think that if the button is of type submit then its fine, please test

**Review by drishu — CHANGES_REQUESTED** [review]
I think for this task we should theme button elements also, I looked into it, to do this we can't override a template, but we can theme from scss
in _all.scss add
```
button.link {
  display: inline;
  margin: 0;
  padding: 0;
  cursor: pointer;
  border: 0;
  background: transparent;
  font-size: 1em;
  text-decoration: underline;
  appearance: none;
}
```
this is needed so that the "Edit summary" button looks ok in create content form (and other forms )

File: templates/patterns/button/pattern-button.html.twig [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
+  'type': type,
+  'outline': outline,
+  'disabled': disabled,
+  'text_nowrap': text_nowrap,
+  'assistive_text': assistive_text,
+  'id': id,
+  'icon': _icon,
+  'icon_position': icon_position,
+  'clean_class': clean_class,
+  'spinner': spinner,
```
Comment (drishu) [inline templates/patterns/button/pattern-button.html.twig:34]: what is this, I don't see it in the field list

File: templates/patterns/button/button.ui_patterns.yml [hunk 2]
```diff
@@ -26,20 +26,56 @@ button:
...
+    icon_position:
+      type: select
+      label: Icon position
+      options:
+        before: before
+        after: after
+    show_spinner:
+      type: boolean
+      label: Show spinner
+    id:
```
Comment (drishu) [inline templates/patterns/button/button.ui_patterns.yml]: this is a reserver word, use soemthing else

**Review by drishu — CHANGES_REQUESTED** [review]

**Review by drishu — CHANGES_REQUESTED** [review]
Looks good, but one more hurdle to pass, the functionality of other patterns was affected because the fields and options changed, please fix:
pattern-collapse.html.twig
pattern-dropdown.html.twig
pattern-modal.html.twig
pattern-popover.html.twig

**Comment by Maxfire** [comment]
> Looks good, but one more hurdle to pass, the functionality of other patterns was affected because the fields and options changed, please fix:
> pattern-collapse.html.twig
> pattern-dropdown.html.twig
> pattern-modal.html.twig
> pattern-popover.html.twig

I checked all and collapse was fixed, also checked different variants, greetings.

File: templates/patterns/button/pattern-button.html.twig [hunk 3]
```diff
@@ -1,18 +1,34 @@
...
+{#
+/**
+ * @file
+ * Button component.
+ */
+#}
+{%  if _icon %}
+  {% set _icon = icon|merge({
+    path: bcl_icon_path
+})  %}
```
Comment (drishu) [inline templates/patterns/button/pattern-button.html.twig]: intendation is not correct, and there is a extra space
```suggestion
  }) %}
```

File: templates/patterns/button/pattern-button.html.twig [hunk 4]
```diff
@@ -1,18 +1,34 @@
...
-  <a{{ attributes.setAttribute('href', url) }}>{{ label }}</a>
-{% else %}
-  {% if disabled %}
-    {% set attributes = attributes.setAttribute('disabled', true) %}
-  {% endif %}
-  <button type="button"{{ attributes }}>{{ label }}</button>
+{#
+/**
+ * @file
+ * Button component.
```
Comment (drishu) [inline templates/patterns/button/pattern-button.html.twig]: * Button pattern.

File: templates/overrides/form/input--submit.html.twig [hunk 5]
```diff
@@ -10,5 +10,13 @@
  * @see template_preprocess_input()
  */
 #}
-{# @TODO Revisit after refactoring the button pattern. #}
-<input{{ attributes.addClass(['btn', 'btn-primary']) }} />{{ children }}
+{% if attributes.hasClass('button--primary') %}
```
Comment (drishu) [inline templates/overrides/form/input--submit.html.twig]: in the login form, the class `button--primary` is not set for the login button, it seems that this class is added per use case and not automatically by a rule.
To solve this issue, we can do something like ``{% elseif attributes['data-drupal-selector'] == 'edit-submit' %}``

**Review by drishu — CHANGES_REQUESTED** [review]

File: templates/patterns/button/pattern-button.html.twig [hunk 6]
```diff
@@ -1,18 +1,34 @@
...
+ */
+#}
+{%  if _icon %}
```
Comment (drishu) [inline templates/patterns/button/pattern-button.html.twig]:
```suggestion
{%  if icon %}
```

File: templates/patterns/button/pattern-button.html.twig [hunk 7]
```diff
@@ -1,18 +1,34 @@
...
+  {% set _icon = icon|merge({
+    path: bcl_icon_path
+})  %}
 {% endif %}
+{%  if show_spinner %}
+  {% set spinner = {
+    'type': "border",
+    'variant': "info",
+    'small': true,
```
Comment (drishu) [inline templates/patterns/button/pattern-button.html.twig:17]: the icon is to close to the text, please add ``attributes: create_attribute().addClass('me-1'),`` like in netlify data tab

File: templates/patterns/button/button.ui_patterns.yml [hunk 8]
```diff
@@ -26,20 +35,60 @@ button:
...
       type: text
       label: Label
       description: The button label
       preview: Submit
+    assistive_text:
+      type: text
+      label: Assistive text
+      description: Additional text hidden with the .visually-hidden class.
+      preview: ''
     url:
```
Comment (drishu) [inline templates/patterns/button/button.ui_patterns.yml]: url doesn't exist

**Review by drishu — CHANGES_REQUESTED** [review]

File: templates/patterns/collapse/pattern-collapse.html.twig [hunk 9]
```diff
@@ -18,13 +18,13 @@
       label: label,
       style: style,
```
Comment (drishu) [inline templates/patterns/collapse/pattern-collapse.html.twig]:
```suggestion
      variant: style,
```

**Review by drishu — CHANGES_REQUESTED** [review]

File: templates/overrides/form/input--submit.html.twig (hunk 5 above)
Comment (Maxfire) [reply templates/overrides/form/input--submit.html.twig]: Ok I'm going to add it as or condition in the same if. thanks.

File: templates/patterns/collapse/pattern-collapse.html.twig (hunk 9 above)
Comment (Maxfire) [reply templates/patterns/collapse/pattern-collapse.html.twig]: this could be a bit confusing but ok.

**Review by abel-santos-corral — APPROVED** [review]

**Review by Maxfire — APPROVED** [review]

**Review by drishu — APPROVED** [review]

**Pull request merged**